File: prime_test/miller_rabin/src/MillerRabinDemo.rb
# Ruby
# 素数判定: ミラーラビン (Miller-Rabin)

def is_prime(target, k=40)
  # 特殊なケースを処理
  return false if target <= 1
  return true if target <= 3
  return false if target % 2 == 0
  
  # n-1 = 2^r * d の形に分解する（dは奇数）
  r, d = 0, target - 1
  while d % 2 == 0
    r += 1
    d /= 2
  end
  
  # k回の試行を実行
  k.times do
    a = rand(2..target-2)
    x = a.pow(d, target)  # a^d mod n を計算
    
    next if x == 1 || x == target - 1
    
    is_composite = true
    (r - 1).times do
      x = x.pow(2, target)
      if x == target - 1
        is_composite = false
        break
      end
    end
    
    return false if is_composite  # 合成数と確定
  end
  
  true  # 素数である可能性が高い
end

def main
  puts "MillerRabin TEST -----> start"

  puts "\nis_prime"
  input_list = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 97, 100, 101, 997]
  input_list.each do |input|
    output = is_prime(input)
    puts "  #{input}: #{output}"
  end
  
  puts "\nMillerRabin TEST <----- end"
end

if __FILE__ == $0
  main
end

File: prime_test/miller_rabin/src/MillerRabinDemo.go
// Go
// 素数判定: ミラーラビン (Miller-Rabin)

package main

import (
	"fmt"
	"math/big"
	"math/rand"
	"time"
)

func isPrime(target int, k int) bool {
	// 特殊なケースを処理
	if target <= 1 {
		return false
	}
	if target <= 3 {
		return true
	}
	if target%2 == 0 {
		return false
	}

	// n-1 = 2^r * d の形に分解する（dは奇数）
	r, d := 0, target-1
	for d%2 == 0 {
		r++
		d /= 2
	}

	// 乱数生成器を初期化
	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))

	// k回の試行を実行
	targetBig := big.NewInt(int64(target))
	for i := 0; i < k; i++ {
		// 2からtarget-2の範囲でランダムな数値aを選択
		a := big.NewInt(0)
		a.Rand(rnd, big.NewInt(int64(target-3)))
		a.Add(a, big.NewInt(2))

		x := big.NewInt(0)
		x.Exp(a, big.NewInt(int64(d)), targetBig) // a^d mod n を計算

		if x.Cmp(big.NewInt(1)) == 0 || x.Cmp(big.NewInt(int64(target-1))) == 0 {
			continue
		}

		testPassed := false
		for j := 0; j < r-1; j++ {
			// x = x^2 mod target
			x.Exp(x, big.NewInt(2), targetBig)
			if x.Cmp(big.NewInt(int64(target-1))) == 0 {
				testPassed = true
				break
			}
		}

		if !testPassed {
			return false // 合成数と確定
		}
	}

	return true // 素数である可能性が高い
}

func main() {
	fmt.Println("MillerRabin TEST -----> start")

	fmt.Println("\nis_prime")
	inputList := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 97, 100, 101, 997}
	for _, input := range inputList {
		output := isPrime(input, 40)
		fmt.Printf("  %d: %t\n", input, output)
	}

	fmt.Println("\nMillerRabin TEST <----- end")
}
